mqtt/mqtt.js

```diff
@@ -1,211 +1,139 @@
+            nats.publish('channel.' + channel, rawMsg);
 
-            console.log(msg);
-
-            console.log(util.inspect(packet, false, null))
-
-            // Pub on NATS
-            nats.publish('channel.' + channel, msg);
+            // Set empty topic for packet so that it won't be published two times.
+            packet.topic = '';
+            publish(0);
         } else {
-            console.log('Publish not authorized');
-            error = 4; // Bad username or password
+            logger.warn("unauthorized publish: %s", err.message);
+            publish(4); // Bad username or password
         }
-        callback(error);
-    });
+    }
 };
 
-// AuthZ SUB
-aedes.authorizeSubscribe = function (client, packet, callback) {
-    // Topics are in the form `mainflux/channels/<channel_id>/messages/senml-json`
-    var channel = packet.topic.split('/')[2];
-    /**
-    * Check if PUB is authorized
-    */
-    var options = {
-        url: config.auth_url + ':' + config.auth_port + '/channels/' + channel + '/access-grant',
-        method: 'GET',
-        headers: {
-            'Content-Type': 'application/json',
-            'Authorization': client.password
-        }
-    };
 
-    request(options, function (err, res) {
-        var error = null;
-        if (res && (res.statusCode === 200)) {
-            console.log('Subscribe authorized OK');
+aedes.authorizeSubscribe = function (client, packet, subscribe) {
+    // Topics are in the form `channels/<channel_id>/messages`
+    var channel = packet.topic.split('/')[1];
+    
+    things.canAccess({
+        token: client.password,
+        chanID: channel
+    }, onAuthorize);
+
+    function onAuthorize(err, res) {
+        if (!err) {
+            logger.info('authorized subscribe');
+            subscribe(null, packet);
         } else {
-            console.log('Subscribe not authorized');
-            error = 4; // Bad username or password
+            logger.warn('unauthorizerd subscribe: %s', err);
+            subscribe(4, packet); // Bad username or password
         }
-        callback(error, packet);
-    });
+    }
 };
 
-// AuthX
-aedes.authenticate = function (client, username, password, callback) {
-    var c = client;
-    var options = {
-        url: config.auth_url + ':' + config.auth_port + '/access-grant',
-        method: 'GET',
-        headers: {
-            'Content-Type': 'application/json',
-            'Authorization': password
-        }
-    };
-    request(options, function (err, res) {
-        var error = null;
-        var success = null;
-        if (res && (res.statusCode === 200)) {
-            // Set MQTT client.id to correspond to Mainflux device UUID
-            c.id = res.headers['x-client-id'];
-            // Store password for future references
-            c.password = password;
-            success = true;
+aedes.authenticate = function (client, username, password, acknowledge) {
+    var pass = password || "",
```

Comment:
I was thinking more about `var pass = (password || "").toString();`. Thus being said, it might be even better idea to apply the same refactoring here as well:

```javascript
// feel free to find more suitable variable names
var token = (password || "").toString(),
    identReq = {value: token},
    onIdentify = function (err, res) { ... };

things.identify(identReq, onIdentify);
```